scriptworker/cot/verify.py

```diff
@@ -0,0 +1,1141 @@
+from scriptworker.utils import format_json, get_hash, load_json, makedirs, match_url_regex, raise_future_exceptions
+from taskcluster.exceptions import TaskclusterFailure
+
+log = logging.getLogger(__name__)
+
+
+# ChainOfTrust {{{1
+class ChainOfTrust(object):
+    """The master Chain of Trust, tracking all the various `LinkOfTrust`s.
+
+    Attributes:
+        context (scriptworker.context.Context): the scriptworker context
+        cot_dir (str): the local path containing this link's artifacts
+        decision_task_id (str): the task_id of self.task's decision task
+        links (list): the list of `LinkOfTrust`s
+        name (str): the name of the task (e.g., signing.decision)
+        task_id (str): the taskId of the task
+        task_type (str): the task type of the task (e.g., decision, build)
+        worker_impl (str): the taskcluster worker class (e.g., docker-worker) of the task
+    """
+    def __init__(self, context, name, task_id=None):
+        self.name = name
+        self.task_type = guess_task_type(name)
+        self.context = context
+        self.task_id = task_id or get_task_id(context.claim_task)
+        self.task = context.task
+        self.worker_impl = guess_worker_impl(self)  # this should be scriptworker
+        self.decision_task_id = get_decision_task_id(self.task)
+        self.links = []
+
+    def dependent_task_ids(self):
+        """Helper method to get all `task_id`s for all `LinkOfTrust` tasks.
+
+        Returns:
+            list: the list of `task_id`s
+        """
+        return [x.task_id for x in self.links]
+
+    def all_tasks(self):
+        """The list of task definitions for all `self.links` + `self`
+
+        Returns:
+            list of dicts: the task definitions for all links + self
+        """
+        return [self.task] + [x.task for x in self.links]
+
+    def is_try(self):
+        """Helper method to determine if any task in the chain is a try task.
+
+        Returns:
+            bool: True if a task is a try task.
+        """
+        result = is_try(self.task)
+        for link in self.links:
+            if link.is_try:
+                result = True
+                break
+        return result
+
+    def get_link(self, task_id):
+        """Get a `LinkOfTrust` by task id.
+
+        Args:
+            task_id (str): the task id to find.
+
+        Returns:
+            LinkOfTrust: the link matching the task id.
+
+        Raises:
+            CoTError: if no `LinkOfTrust` matches.
+        """
+        links = [x for x in self.links if x.task_id == task_id]
+        if len(links) != 1:
+            raise CoTError("No single Link matches task_id {}!\n{}".format(task_id, self.dependent_task_ids()))
+        return links[0]
+
+
+# LinkOfTrust {{{1
+class LinkOfTrust(object):
+    """Each LinkOfTrust represents a task in the Chain of Trust and its status.
+
+    Attributes:
+        chain (ChainOfTrust): the ChainOfTrust object.
```

Comment:
We can.  The difference is an `AttributeError` when we reference `link.chain` or an `AttributeError` when we reference `link.chain.$attr`.  I preferred the earlier `AttributeError`, BUT it looks like we never use `link.chain` so I'm going to just tear it out :)